### PR TITLE
Add support for cross-site communication using Openshift Routes

### PR DESCRIPTION
**Pull request by pruivo**
Enable Openshift Route for cross-site communication. 

Currently, I'm unable to add tests for CI. :/

### CODE REVIEW

File: controllers/xsite_config.go [hunk 1]
```diff
@@ -112,18 +139,40 @@ func appendRemoteLocation(ctx context.Context, infinispan *ispnv1.Infinispan, re
...
+	if err == nil {
+		host, port, err = getCrossSiteServiceHostPort(siteService, remoteKubernetes, logger, eventRec, "XSiteRemoteServiceUnsupported", ctx)
+		if err != nil {
+			logger.Error(err, "error retrieving remote x-site service information")
+			return err
+		}
+	} else {
+		// Try searching for the Route
+		siteRoute := &routev1.Route{}
+		err = remoteKubernetes.Client.Get(context.TODO(), types.NamespacedName{Name: remoteRouteName, Namespace: remoteNamespace}, siteRoute)
```
Comment (ryanemerson) [inline controllers/xsite_config.go:169]: We need to use the `ctx` provided as a parameter.

File: Makefile [hunk 2]
```diff
@@ -134,7 +134,7 @@ generate: controller-gen rice
 .PHONY: operator-build
 ## Build the operator image
 operator-build: manager
-	$(CONTAINER_TOOL) build -t $(IMG) .
+	$(CONTAINER_TOOL) build --build-arg VERSION=$(VERSION) -t $(IMG) .
```
Comment (ryanemerson) [inline Makefile:137]: Why is this necessary?

**Review by ryanemerson — CHANGES_REQUESTED** [review]

**Comment by ryanemerson** [comment]
> Currently, I'm unable to add tests for CI. :/

You could add a test that works with Openshift. We can then perform a quick check at the start of the test method to determine if the k8s flavour supports Routes and if not calls `t.Skip()`.

File: Makefile (hunk 2 above)
Comment (pruivo) [reply Makefile:137]: There is a `VERSION` argument in the Dockerfile that is not set and it does not have a default value. 
Isn't an issue?

Podman does not like it
```
$ make operator-build CONTAINER_TOOL=podman
/home/pedro/Work/infinispan-operator/bin/controller-gen object paths="./..."
/home/pedro/Work/infinispan-operator/bin/rice embed-go -i controllers/dependencies.go -i controllers/grafana.go
find . -type f -name 'rice-box.go' -exec sed -i "s|time.Unix(.*, 0)|time.Unix(1620137619, 0)|" {} \;
go fmt ./...
go vet ./...
go build -o bin/manager main.go
podman build -t quay.io/infinispan/operator:latest .
WARN[0001] missing "VERSION" build argument. Try adding "--build-arg VERSION=<VALUE>" to the command line 
```

Comment (pruivo) [reply Makefile:137]: actually, `VERSION` is set in the base image.
Adding `RUN echo ${VERSION}` I got (with docker)
```
Step 1/11 : FROM registry.access.redhat.com/ubi8/go-toolset:1.15.14 AS build
...
Step 3/11 : RUN echo ${VERSION}
 ---> Running in 6aeb06d29666
1.15.14
```
Is it intentional?

Comment (ryanemerson) [reply Makefile:137]: > There is a `VERSION` argument in the Dockerfile that is not set and it does not have a default value.
> Isn't an issue?

No it is, good spot :+1: 

> actually, `VERSION` is set in the base image. Adding `RUN echo ${VERSION}` I got (with docker)
> Is it intentional?

Nope :sweat_smile:

File: controllers/xsite_config.go (hunk 1 above)
Comment (pruivo) [reply controllers/xsite_config.go]: fixed!

**Comment by pruivo** [comment]
> We can then perform a quick check at the start of the test method to determine if the k8s flavour supports Routes

I need time to investigate how to do it :)

**Comment by ryanemerson** [comment]
> I need time to investigate how to do it :)

We already have a method to determine if a GVK is supported on the k8s platform: https://github.com/infinispan/infinispan-operator/blob/main/pkg/kubernetes/kubernetes.go#L72-L92

**Comment by ryanemerson** [comment]
@pruivo I knew I had already implemented something similar for the testsuite, but couldn't find it ... you can use the `tutils.SkipForExposeType` to ensure a test is only executed when the `EXPOSE_SERVICE_TYPE=Route`.

https://github.com/ryanemerson/infinispan-operator/commit/c8219e311f827d226a5a34c53427441e398f3833

**Comment by pruivo** [comment]
it looks like that commit is not available in the operator. Do you mind if I cherry-pick it to this PR?

**Comment by ryanemerson** [comment]
@pruivo That explains why I couldn't find the code :upside_down_face:.  Go for it!

File: config/crd/bases/infinispan.org_infinispans.yaml [hunk 3]
```diff
@@ -921,13 +921,18 @@ spec:
                               port:
                                 format: int32
                                 type: integer
+                              routerHostName:
```
Comment (Crumby) [inline config/crd/bases/infinispan.org_infinispans.yaml:924]: Here is inconsistency between field name and the description. While the field is `routerHostName` the description mentions `RouteHostName` (route vs router), probably should be `route`

File: controllers/infinispan_service_controller.go [hunk 4]
```diff
@@ -538,6 +554,39 @@ func computeSiteService(ispn *ispnv1.Infinispan) *corev1.Service {
...
+			APIVersion: "route.openshift.io/v1",
+			Kind:       "Route",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ispn.GetSiteRouteName(),
+			Namespace: ispn.Namespace,
+			Labels:    LabelsResource(ispn.Name, "infinispan-route-xsite"),
+		},
+		Spec: routev1.RouteSpec{
+			Host: host,
```
Comment (Crumby) [inline controllers/infinispan_service_controller.go:570]: What if the `host` is empty?

File: test/e2e/xsite/xsite_test.go [hunk 5]
```diff
@@ -231,6 +231,16 @@ func TestDefaultTLSLoadBalancerWithPort(t *testing.T) {
 	testCrossSiteView(t, true, ispnv1.CrossSiteSchemeTypeOpenShift, ispnv1.CrossSiteExposeTypeLoadBalancer, 1443, 1, DefaultTLS, nil)
 }
 
+func TestCrossSiteViewOpenshiftRoute(t *testing.T) {
+	tutils.SkipForExposeType(t, ispnv1.ExposeTypeNodePort, ispnv1.ExposeTypeLoadBalancer)
```
Comment (Crumby) [inline test/e2e/xsite/xsite_test.go:235]: Maybe skipping based on whether `Route` API is available would be more consistent with the rest of the test rather then on test suite configuration? (We are not skipping LoadBalancer tests just because NodePort expose type is configured)

File: test/e2e/xsite/xsite_test.go [hunk 6]
```diff
@@ -231,6 +231,16 @@ func TestDefaultTLSLoadBalancerWithPort(t *testing.T) {
...
+	testCrossSiteView(t, true, ispnv1.CrossSiteSchemeTypeOpenShift, ispnv1.CrossSiteExposeTypeRoute, 0, 1, NoTLS, nil)
```
Comment (Crumby) [inline test/e2e/xsite/xsite_test.go:236]: `NoTLS` option is confusing. Cross site over Route won't work without TLS

File: test/e2e/xsite/xsite_test.go [hunk 7]
```diff
@@ -231,6 +231,16 @@ func TestDefaultTLSLoadBalancerWithPort(t *testing.T) {
 	testCrossSiteView(t, true, ispnv1.CrossSiteSchemeTypeOpenShift, ispnv1.CrossSiteExposeTypeLoadBalancer, 1443, 1, DefaultTLS, nil)
 }
 
+func TestCrossSiteViewOpenshiftRoute(t *testing.T) {
```
Comment (Crumby) [inline test/e2e/xsite/xsite_test.go:234]: Having internal version of the test for smoke testing on a single cluster would be nice

Comment (pruivo) [reply test/e2e/xsite/xsite_test.go:234]: that is useless. Cross-site in a single cluster uses the cluster IP directly and not the Routes (same for NodePort & LoadBalancer)

File: test/e2e/xsite/xsite_test.go (hunk 6 above)
Comment (pruivo) [reply test/e2e/xsite/xsite_test.go:236]: I'm going to remove this test. It does not make sense to me :rofl:

File: test/e2e/xsite/xsite_test.go (hunk 5 above)
Comment (pruivo) [reply test/e2e/xsite/xsite_test.go:235]: any pointers how to do that?

Comment (ryanemerson) [reply test/e2e/xsite/xsite_test.go]: We can make use of the `IsGroupVersionSupported` method like [here](https://github.com/infinispan/infinispan-operator/blob/main/test/e2e/utils/common.go#L216) to see if it's supported and if not call skip on the test.

File: test/e2e/xsite/xsite_test.go (hunk 7 above)
Comment (Crumby) [reply test/e2e/xsite/xsite_test.go]: Right, I incorrectly though we have "single cluster" version of the tests for NodePort and LoadBalancer too.

File: controllers/infinispan_service_controller.go (hunk 4 above)
Comment (Crumby) [reply controllers/infinispan_service_controller.go:576]: Right, I was more wondering whether OpenShift won't complain if you send a request with explicit `host: nil`, compared to "not mentioning" `host` at all. But if it works then good, it's definitely nicer then having if block.

**Comment by pruivo** [comment]
Updated!
* Fixed `TestCrossSiteViewKubernetesLoadBalancer`
* Tested the skip() method
```
=== RUN   TestDefaultTLSOpenshiftRoute
    xsite_test.go:297: Route not available. Skipping test
--- SKIP: TestDefaultTLSOpenshiftRoute (6.09s)
```

**Comment by pruivo** [comment]
green :partying_face:

**Comment by ryanemerson** [comment]
Thanks @pruivo
